File: src/services/ai/prompt.js
// Construye el system prompt para el bot de Seguros Carguill

function buildSystemPrompt(cfg, faqs, slotsInfo) {
  const c = cfg.contexto || {};
  const partes = [];

  // --- IDENTIDAD ---
  const empresa = c.empresa || 'Seguros Carguill';
  const ciudad  = c.ciudad  || 'San Pedro Garza García, NL';
  partes.push(
    `Eres el asistente virtual de ${empresa}, un broker de seguros ubicado en ${ciudad}, México. ` +
    `Tu nombre es Carguill. Responde siempre en español, de forma clara y concisa.`
  );

  // --- ROL Y MISIÓN ---
  partes.push(
    `Tu rol es orientar y asesorar a los clientes que buscan un seguro. ` +
    `NO eres un vendedor directo: tu función es entender la necesidad del cliente, ` +
    `recopilar la información relevante y conectarlo con un asesor humano para la cotización formal. ` +
    `Trabajamos con más de 25 aseguradoras en México para encontrar la mejor opción para cada cliente.`
  );

  // --- SEGUROS DISPONIBLES ---
  const seguros = Array.isArray(c.seguros) && c.seguros.length > 0
    ? c.seguros
    : ['Vida', 'Gastos Médicos Mayores', 'Auto', 'Daños', 'Viaje', 'Mascotas'];
  partes.push(`Seguros que ofrecemos:\n${seguros.map(s => `• ${s}`).join('\n')}`);

  // --- ASEGURADORAS ---
  if (c.aseguradoras) {
    partes.push(`Aseguradoras con las que trabajamos: ${c.aseguradoras}.`);
  }

  // --- FLUJO DE ATENCIÓN POR TIPO DE SEGURO ---
  partes.push(`
FLUJO DE ATENCIÓN (síguelo en orden):

1. Saluda al cliente y pregunta qué tipo de seguro le interesa o qué necesita.
2. Una vez detectado el tipo de seguro, llama a "registrar_interes" para registrarlo.
3. Recopila los datos básicos según el tipo:
   • Vida: edad del titular, si fuma, monto de cobertura deseado.
   • Gastos Médicos: edad, número de personas a asegurar, ¿tiene condiciones preexistentes?
   • Auto: año, marca y modelo del vehículo, uso (personal/comercial), ¿amplia o limitada?
   • Daños: tipo de bien (casa, negocio, equipo), ubicación, valor aproximado.
   • Viaje: destino, fechas, número de viajeros, ¿tiene cobertura médica incluida?
   • Mascotas: especie, raza, edad de la mascota.
4. Con esa información, ofrece al cliente agendar una llamada o cita con un asesor para la cotización formal.
   Si el cliente acepta, usa "agendar_cita".
5. Si el cliente pregunta precios específicos, dile que los precios dependen de sus datos y que un asesor
   le enviará cotizaciones comparativas de varias aseguradoras. No inventes precios.
6. Si el cliente pide hablar con una persona o la consulta es muy compleja, usa "escalar_a_agente".`.trim()
  );

  // --- HORARIO ---
  if (c.horario) {
    partes.push(`Horario de atención: ${c.horario}.`);
  }

  // --- TONO ---
  const tonos = {
    formal:   'Responde en tono formal y profesional.',
    amigable: 'Responde en tono amigable y cercano.',
    ambos:    'Responde en tono formal pero amigable y cálido.',
  };
  if (c.tono && tonos[c.tono]) partes.push(tonos[c.tono]);

  // --- BIENVENIDA ---
  if (c.bienvenida) {
    partes.push(`Cuando un cliente escriba por primera vez salúdalo así: "${c.bienvenida}"`);
  }

  // --- RESTRICCIONES ---
  partes.push(
    `RESTRICCIONES IMPORTANTES:\n` +
    `• Nunca inventes precios, primas ni coberturas específicas.\n` +
    `• Nunca prometas aprobación de una póliza.\n` +
    `• No proporciones datos de clientes de terceros.\n` +
    `• Si no sabes algo, dilo honestamente y ofrece conectar con un asesor.\n` +
    (c.restricciones ? `• ${c.restricciones}` : '')
  );

  // --- INSTRUCCIONES ADICIONALES ---
  if ((cfg.instrucciones || '').trim()) {
    partes.push(cfg.instrucciones.trim());
  }

  // --- BASE DE CONOCIMIENTO (FAQs) ---
  if (faqs && faqs.length > 0) {
    const faqTexto = faqs.map(f => `P: ${f.pregunta}\nR: ${f.respuesta}`).join('\n\n');
    partes.push(`BASE DE CONOCIMIENTO:\n${faqTexto}`);
  }

  // --- SLOTS DE CALENDARIO ---
  if (slotsInfo && slotsInfo.texto) {
    partes.push(
      slotsInfo.texto +
      '\nSi el cliente quiere agendar, usa la función "agendar_cita" con el índice del slot elegido (0-based).'
    );
  }

  return partes.join('\n\n');
}

function buildMessages(systemPrompt, mensajes) {
  const messages = [{ role: 'system', content: systemPrompt }];
  for (const m of [...mensajes].reverse()) {
    messages.push({
      role: m.autor === 'cliente' ? 'user' : 'assistant',
      content: m.contenido,
    });
  }
  return messages;
}

module.exports = { buildSystemPrompt, buildMessages };

File: app/services/ai/prompt.py
def build_system_prompt(cfg: dict, faqs: list, slots_info: dict) -> str:
    c = cfg.get("contexto") or {}
    partes = []

    # ── Identidad: 100% desde la configuración del panel (sin nada hardcodeado de aseguradora) ──
    nombre_bot = (c.get("bot_nombre") or "Asistente").strip()
    empresa = (c.get("empresa") or "la empresa").strip()
    ciudad = (c.get("ciudad") or "").strip()
    ubic = f", ubicada en {ciudad}, México" if ciudad else ""
    partes.append(
        f"Eres {nombre_bot}, el asistente virtual de {empresa}{ubic}. "
        "Responde siempre en español, de forma clara y concisa."
    )

    # Descripción/rol del negocio: lo define el panel. Si no hay, un rol neutro mínimo.
    descripcion = (c.get("descripcion") or "").strip()
    if descripcion:
        partes.append(descripcion)
    else:
        partes.append(
            f"Tu función es atender a los clientes de {empresa}: entender lo que necesitan, "
            "resolver sus dudas con la información disponible y conectarlos con un asesor humano "
            "cuando haga falta."
        )

    if c.get("seguros"):
        partes.append("Productos/servicios que ofrecemos:\n" + "\n".join(f"• {s}" for s in c["seguros"]))

    if c.get("aseguradoras"):
        partes.append(f"Aseguradoras con las que trabajamos: {c['aseguradoras']}.")

    # Flujo de atención: lo define el panel (campo "proceso"). Ya no hay flujo de aseguradora fijo.
    if c.get("proceso"):
        partes.append("CÓMO DEBES ATENDER (síguelo en orden):\n" + c["proceso"].strip())

    # Las funciones siguen disponibles con cualquier flujo.
    partes.append(
        "Tienes funciones para registrar el interés del cliente, escalar la conversación a un "
        "asesor humano y agendar una cita; úsalas cuando corresponda."
    )

    # Datos que el bot SIEMPRE debe recopilar
    if c.get("datos_obligatorios"):
        partes.append(
            "Antes de agendar o escalar, asegúrate de tener estos datos del cliente:\n"
            + c["datos_obligatorios"].strip()
        )

    # Datos de la oficina
    oficina = []
    if c.get("direccion"): oficina.append(f"Dirección: {c['direccion']}")
    if c.get("telefono"): oficina.append(f"Teléfono: {c['telefono']}")
    if c.get("web"): oficina.append(f"Sitio web: {c['web']}")
    if oficina:
        partes.append("Datos de la oficina:\n" + "\n".join(f"• {o}" for o in oficina))

    if c.get("horario"):
        partes.append(f"Horario de atención: {c['horario']}.")

    tonos = {
        "formal":   "Responde en tono formal y profesional.",
        "amigable": "Responde en tono amigable y cercano.",
        "ambos":    "Responde en tono formal pero amigable y cálido.",
    }
    if c.get("tono") and c["tono"] in tonos:
        partes.append(tonos[c["tono"]])

    if c.get("bienvenida"):
        partes.append(f'Cuando un cliente escriba por primera vez salúdalo así: "{c["bienvenida"]}"')

    restricciones_extra = f"\n• {c['restricciones']}" if c.get("restricciones") else ""
    politica = f"\n• {c['politica_precios']}" if c.get("politica_precios") else ""
    prohibidos = f"\n• Temas prohibidos (no respondas sobre esto, escala a un asesor): {c['temas_prohibidos']}" if c.get("temas_prohibidos") else ""
    partes.append(
        "RESTRICCIONES IMPORTANTES:\n"
        "• NO inventes ni supongas información. Si algo NO está en esta configuración ni en la "
        "BASE DE CONOCIMIENTO, dilo con honestidad y ofrece conectar con un asesor humano. "
        "Prefiere decir \"no tengo ese dato, te conecto con un asesor\" antes que arriesgar una respuesta.\n"
        "• No inventes precios, cifras, fechas ni hagas promesas.\n"
        "• No proporciones datos personales de terceros.\n"
        f"• No te desvíes a temas ajenos a {empresa}; si insisten, ofrece pasar con un asesor.{politica}{prohibidos}{restricciones_extra}"
    )

    if cfg.get("instrucciones", "").strip():
        partes.append(cfg["instrucciones"].strip())

    if faqs:
        faq_texto = "\n\n".join(f"P: {f['pregunta']}\nR: {f['respuesta']}" for f in faqs)
        partes.append(f"BASE DE CONOCIMIENTO:\n{faq_texto}")

    if slots_info and slots_info.get("texto"):
        partes.append(
            slots_info["texto"] +
            '\nSi el cliente quiere agendar, usa la función "agendar_cita" con el índice del slot elegido (0-based).'
        )

    # Límite de longitud y firma
    if c.get("max_palabras"):
        partes.append(f"Responde de forma breve: máximo {c['max_palabras']} palabras por mensaje.")
    if c.get("firma"):
        partes.append(f'Cierra tus mensajes con: "{c["firma"]}"')

    # MODO ESTRICTO: el bot responde SOLO con lo configurado; si no, pasa a un humano.
    if c.get("modo_estricto"):
        handoff = (c.get("mensaje_handoff") or
                   "Excelente pregunta. Déjame conectarte con un asesor que te dará el detalle exacto. 🙌")
        partes.append(
            "MODO ESTRICTO (OBLIGATORIO): Responde ÚNICAMENTE con base en la información de esta "
            "configuración, las RESTRICCIONES y la BASE DE CONOCIMIENTO. Si la información necesaria "
            "para responder NO está aquí, NO la inventes, NO supongas ni uses conocimiento externo: "
            'llama a la función "escalar_a_agente" y responde EXACTAMENTE con este texto: '
            f'"{handoff}"'
        )

    return "\n\n".join(partes)


def build_messages(system_prompt: str, mensajes: list) -> list:
    messages = [{"role": "system", "content": system_prompt}]
    for m in reversed(mensajes):
        role = "user" if m["autor"] == "cliente" else "assistant"
        messages.append({"role": role, "content": m["contenido"]})
    return messages
